oxidizer/src/db/connections.rs

```diff
@@ -0,0 +1,126 @@
+use async_trait::async_trait;
+use tokio_postgres::{Client, Config, NoTls};
+
+use crate::Error;
+
+#[async_trait]
+pub trait ConnectionProvider: Send + Sync + 'static {
+    async fn connect(&self) -> Result<Client, tokio_postgres::Error>;
+}
+
+pub fn create_connection_provider(
+    config: Config,
+    ca_file: Option<&str>,
+) -> Result<Box<dyn ConnectionProvider>, Error> {
+    let prov: Box<dyn ConnectionProvider> = if let Some(ca_file) = ca_file {
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "tls-openssl")] {
+                tls_openssl::new_tls(config, ca_file)?
+            } else if #[cfg(feature = "tls-rustls")] {
+                tls_rustls::new_tls(config, ca_file)?
+            }else {
+                eprintln!("[WARN] no TLS provider found, reverting to unencrypted connection");
+                no_tls(config)
+            }
+        }
+    } else {
+        no_tls(config)
+    };
+
+    Ok(prov)
+}
+
+struct NoTlsConnectionProvider {
+    config: Config,
+}
+
+#[async_trait]
+impl ConnectionProvider for NoTlsConnectionProvider {
+    async fn connect(&self) -> Result<Client, tokio_postgres::Error> {
+        let (client, conn) = self.config.connect(NoTls).await?;
+        mobc::spawn(conn);
+        Ok(client)
+    }
+}
+
+fn no_tls(config: Config) -> Box<dyn ConnectionProvider> {
+    Box::new(NoTlsConnectionProvider { config })
+}
+
+#[cfg(feature = "tls-openssl")]
+mod tls_openssl {
+    use super::*;
+
+    use openssl::ssl::{SslConnector, SslMethod, SslOptions};
+
+    use postgres_openssl::MakeTlsConnector;
+    use tokio_postgres::{Client, Config};
+
+    struct OpensslConnectionProvider {
+        config: Config,
+        tls: MakeTlsConnector,
+    }
+
+    #[async_trait]
+    impl ConnectionProvider for OpensslConnectionProvider {
+        async fn connect(&self) -> Result<Client, tokio_postgres::Error> {
+            let (client, conn) = self.config.connect(self.tls.clone()).await?;
+            mobc::spawn(conn);
+            Ok(client)
+        }
+    }
+
+    pub fn new_tls(config: Config, ca_file: &str) -> Result<Box<dyn ConnectionProvider>, Error> {
+        let mut builder = SslConnector::builder(SslMethod::tls()).map_err(Error::OpensslError)?;
```

Comment:
Can we make this fn async please? And I think this builder is blocking right? See if it's possible to make it async with tokio. Thanks!